### PR TITLE
[refactor] Add bucket validation util functions

### PR DESCRIPTION
**Pull request by envestcc**
# Description

introduce batch of util functions to validate a bucket, reducing redundant code

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by millken — APPROVED** [review]

File: action/protocol/staking/bucket_validation.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		return &handleError{
+			err:           errors.New("bucket amount is unsufficient"),
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketAmount,
+		}
+	}
+	return nil
+}
+
+func validateBucketStake(bucket *VoteBucket, isStaked bool) ReceiptError {
+	if bucket.isUnstaked() == isStaked {
```
Comment (dustinxie) [inline action/protocol/staking/bucket_validation.go:34]: so it returns error for `isStaked` = both true and false?

File: action/protocol/staking/bucket_validation.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		}
+		return &handleError{
+			err:           err,
+			failureStatus: iotextypes.ReceiptStatus_ErrInvalidBucketType,
+		}
+	}
+	return nil
+}
+
+func validateBucketCandidate(bucket *VoteBucket, candidate address.Address) ReceiptError {
```
Comment (dustinxie) [inline action/protocol/staking/bucket_validation.go:47]: where is this needed?

File: action/protocol/staking/bucket_validation_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		bktIdx, err := csm.putBucketAndIndex(bkt)
+		r.NoError(err)
+		blkHeight := uint64(10)
+		// not endorsed bucket
+		r.Nil(validateBucketEndorsement(esm, bkt, false, blkHeight))
+		r.ErrorContains(validateBucketEndorsement(esm, bkt, true, blkHeight), "bucket is not endorsed")
+		// endorsed bucket
+		r.NoError(esm.Put(bktIdx, &Endorsement{ExpireHeight: endorsementNotExpireHeight}))
+		r.Nil(validateBucketEndorsement(esm, bkt, (true), blkHeight))
+		r.ErrorContains(validateBucketEndorsement(esm, bkt, (false), blkHeight), "bucket is already endorsed")
```
Comment (dustinxie) [inline action/protocol/staking/bucket_validation_test.go:111]: `false`, not `(false)`

File: action/protocol/staking/bucket_validation_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		// unendorsing bucket
+		r.NoError(esm.Put(bktIdx, &Endorsement{ExpireHeight: blkHeight + 1}))
+		r.Nil(validateBucketEndorsement(esm, bkt, (true), blkHeight))
+		r.ErrorContains(validateBucketEndorsement(esm, bkt, (false), blkHeight), "bucket is already endorsed")
```
Comment (dustinxie) [inline action/protocol/staking/bucket_validation_test.go:115]: same here

File: action/protocol/staking/bucket_validation_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		// endorse expired bucket
+		r.NoError(esm.Put(bktIdx, &Endorsement{ExpireHeight: blkHeight}))
+		r.Nil(validateBucketEndorsement(esm, bkt, (false), blkHeight))
+		r.ErrorContains(validateBucketEndorsement(esm, bkt, (true), blkHeight), "bucket is not endorsed")
```
Comment (dustinxie) [inline action/protocol/staking/bucket_validation_test.go:119]: and these 2

**Review by dustinxie — APPROVED** [review]

**Pull request merged**